AWS: Add progressive multipart upload to S3FileIO

danielcweeks opened the pull request:
Add progressive upload to S3OutputStream using multipart upload.

A few key changes are:

1. The output stream will switch from single PUT based upload to multipart when the multipart size * threshold is reached.
2. Staging files will be rotated and uploaded async as output written
3. Staging files will be deleted when the upload completes to reduce local disk requirements
4. An attempt will be made to abort the upload upon failure, but this is not guaranteed (Bucket lifecycle rules should be implemented to ensure data is cleaned up.)

danielcweeks commented:
@jackye1995 it would be great to get your thoughts on this approach.  Still needs some work and lots more testing.

jackye1995 commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,18 +168,105 @@ public void close() throws IOException {
     } finally {
-      if (!stagingFile.delete()) {
-        LOG.warn("Could not delete temporary file: {}", stagingFile);
+      stagingFiles.forEach(f -> {
```
Can we also delete the temp file progressively instead of deleting everything after close() is called?

danielcweeks replied:
Good catch! That was the intent and I think at some point I had it in the async upload task.  Should be easy to add back there after the upload is successful.

rdblue replied:
Also, we can add reliability by using `Tasks.foreach`. That would allow retries, failure logging, and deleting each file even if another has failed.

danielcweeks replied:
Yeah, let me try switching over to Tasks.  I was having trouble with figuring out how to properly handle the exception cases with CompletableFuture without swallowing the failures.

danielcweeks replied:
@rdblue looking at `Tasks.foreach`, I'm not sure how well it fits in this case.  It looks like `Tasks` assumes that you know the full units of work and will submit them in parallel and block.  In this case, we are progressively adding more work and then need to block across all of those at the end to complete the multipart upload.  Now, I could wrap the `Tasks.foreach` call in future, but that seems like we're just wrapping one async framework with another.

I think we can handle the concerns of deleting the files, aborting the upload, and logging with `CompletableFuture` framework and retries should be left to the S3Client (trying to avoid retries on top of retries).

Thoughts?

rdblue replied:
Are you saying that `stagingFiles` is changing as this loops through it?

I think that when we try to revert the uploads, we should know that no more uploads are happening concurrently so that the tasks can run. Maybe I'm misunderstanding your point though?

danielcweeks replied:
I think I'm the one who got confused here.  I though you were suggesting to move a different section to using `Tasks`.  For the actual cleanup, Tasks will probably work.  I'll take another look.

danielcweeks replied:
Ok, so there are actually two areas to resolve here (which was my original confusion).

1.  Delete the staging file when an upload is success full, which I added
2.  Convert the cleanup during abort to use Tasks framework.

jackye1995 commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,18 +168,105 @@ public void close() throws IOException {
+              .contentLength(f.length())
+              .build();
+
+          CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
```
If we want to make it async, can we directly leverage `S3AsyncClient`?

danielcweeks replied:
That's a really good question.  Since it's the entire client that changes (right?) this would also require changes in the input stream.  Someone on the original thread commented they had problems with the async implementation, so I had steered away from it.

jackye1995 replied:
Which original thread do you mean? Do you have a link?

johnclara replied:
we use the StsAssumeRoleCredentialsProvider to keep a fresh session when running on prem. right now this doesnt' support StsAsyncClient and only supports StsClient.

Also the aws sdk provided nio client doesn't look like it supports proxies with basic auth (needed for us to leave our dc). The CRT one does, but it seems like it's still young?

https://github.com/aws/aws-sdk-java-v2/blob/master/services/sts/src/main/java/software/amazon/awssdk/services/sts/auth/StsAssumeRoleCredentialsProvider.java

danielcweeks replied:
@jackye1995 the comment I was referring to about the async client was this: https://github.com/apache/iceberg/pull/1573#discussion_r502725465

Overall, I'm not sure we benefit much from the async client in this case because everything around is still synchronous.  I feel like if we move to a vectored-io implementation for read, that might be a better time to explore switching to async.

jackye1995 replied:
Yeah I know there are some limitations right now for the NioNettyAsyncHttpClient. In this case let's just use the synchronous client then. Using `Tasks` suggested by Ryan sounds like a good idea.

jackye1995 reviewed (commented):
Thank you! Looks good to me, the two big questions I have regarding the upload logic are (1) can we progressively delete staging files, and (2) can we leverage `S3AsyncClient` for async upload.

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -19,42 +19,83 @@
+  private String multipartUploadId;
+  private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
+  private final int multiPartSize;
+  private final double multiPartThresholdFactor;
```
I think I went with factor because it makes it easier to tune the part size.  If you want smaller (or larger) part sizes, you need to adjust the part size and make sure the threshold size is appropriately sized.  With the factor, you just need to adjust the part size and the default factor will typically work well.  (Though now that you point this out we probably need more validation either way).

danielcweeks replied:
I updated the code to use a threshold size, but still expose the factor as the user facing property to adjust.

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/BaseS3File.java:
```diff
@@ -29,10 +31,18 @@
   BaseS3File(S3Client client, S3URI uri) {
     this.client = client;
     this.uri = uri;
+    this.properties = Collections.emptyMap();
```
I actually like what you have with the `AwsCatalogProperties`, but maybe we could use that as utility for translating to/from the catalog properties.  I feel like there are going to be a lot of these configs so we probably don't want to pull each one out individually.  The catalog properties are also hard to deal with becuase it's just `Map<String,String>` which makes defaults/casting harder.  I'm thinking of something like `awsCatalogProperties = AwsCatalogProperties.from(properties)` and then we can just use `awsCatalogProperties.multipartSize()` (for example).

rdblue replied:
I like the idea of getting rid of the map as soon as it is used. That way we don't have to serialize it to tasks.

danielcweeks replied:
I've converted these over to using AwsProperties

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -19,42 +19,105 @@
-  S3OutputStream(S3Client s3, S3URI location) throws IOException {
+  S3OutputStream(Map<String, String> properties, S3Client s3, S3URI location) throws IOException {
+    synchronized (this) {
+      if (executorService == null) {
```
If this check were done outside of `synchronized` as well, then we wouldn't need to get the lock each time a file is opened. There isn't a huge chance of lock contention, but it is always nice to avoid locking when possible.

danielcweeks replied:
I'll add the double-check, but like you said, this really isn't a critical section (and I believe the the stack grab we do below is probably much more expensive and likely to cause some sort of locking).

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -19,42 +19,105 @@
+      if (executorService == null) {
+        executorService = MoreExecutors.getExitingExecutorService(
+            (ThreadPoolExecutor) Executors.newFixedThreadPool(
+                Integer.parseInt(properties.getOrDefault(UPLOAD_POOL_SIZE,
```
Iceberg has `PropertyUtil.propertyAsInt` for this situation.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,17 +190,105 @@ public void close() throws IOException {
+      long contentLength = stagingFiles.stream().mapToLong(File::length).sum();
+      InputStream contentStream = new BufferedInputStream(stagingFiles.stream()
+          .map(S3OutputStream::uncheckedInputStream)
+          .reduce(SequenceInputStream::new).orElseGet(() -> new ByteArrayInputStream(new byte[0])));
```
Nit: Could you move `orElseGet` to a new line?

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,17 +190,105 @@ public void close() throws IOException {
     try {
       stream.close();
 
+      completeUploads();
```
It looks like uploading the last part depends on the line above where `currentStagingFile` is set to `null`. Otherwise, the current part is ignored in `uploadParts`. This leads to slightly confusing behavior for `completeUploads` because it won't include the last part unless that assumption is met.

What about making `completeUploads` ensure the current stream is closed and null using a `closeCurrent` method?

danielcweeks replied:
So, I changed the logic a bit to address this.  It now does not require the current file to be null, only requires that the stream is closed and I added a check to enforce that.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,17 +190,105 @@ public void close() throws IOException {
+    } catch (CompletionException e) {
+      abortUpload();
+      throw new IOException("Multipart upload failed for upload id: " + multipartUploadId, e);
+    }
```
Here's another place where we can add reliability using `Tasks`. That has a callback for failure that can fail with its own exception, which will be suppressed instead of thrown in place of the `CompletionException`.

rdblue commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -69,14 +132,54 @@ public void flush() throws IOException {
+    if (stream.getCount() >= multiPartSize) {
+      newStream();
+      uploadParts();
+    }
```
When this is used, the class will never switch to progressive upload. Is that intentional?

danielcweeks replied:
Yeah, that was just an oversight, it should initiate as well.

johnclara commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -69,14 +132,54 @@ public void flush() throws IOException {
+    // and continue to write new staging files if the write is
+    // larger than the part size.
+    while (stream.getCount() + remaining > multiPartSize) {
+      int writeSize = multiPartSize - (int) stream.getCount();
```
maybe also try to initialize the multipart upload up here so that the individual parts can get kicked off immediately during the loop?

(That currently returns immediately if this is the first chunk right?)

danielcweeks replied:
I feel like this is really unlikely to be an issue.  The loop is mostly for completeness, but for you to hit a second iteration, the write size would have to be 2x the page size and given this is internally wrapped in a buffered output stream, the part size would need to be less than 4K (less than half the output buffer size).  

I think your right that it wouldn't start uploading parts until after the loop, but I just don't feel like it's worth the extra complexity.

johnclara replied:
Makes sense!

johnclara commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -87,17 +190,105 @@ public void close() throws IOException {
+  }
+
+  private void initializeMultiPartUpload() {
+    multipartUploadId = s3.createMultipartUpload(CreateMultipartUploadRequest.builder()
```
s3a supports canned acls for s3 requests, not sure if this should also support them?

My team had to use them once for writing from within account A to a bucket owned by bucket B

We also had to add it to rdblue/s3committer:
https://github.com/rdblue/s3committer/blob/master/src/main/java/com/netflix/bdp/s3/S3Util.java#L73
```
    InitiateMultipartUploadRequest initiateMultipartUploadRequest = new InitiateMultipartUploadRequest(bucket, key)
          .withCannedACL(CannedAccessControlList.BucketOwnerFullControl);
    InitiateMultipartUploadResult initiate = client.initiateMultipartUpload(initiateMultipartUploadRequest);
```

danielcweeks replied:
Interesting, I didn't even know about canned acls.  Looks like the builder does support it (via `.acl()`), so that's probably  another thing we can plumb through.

johnclara replied:
For sure, it should probably go in another PR.

Should I file an issue? Or is the issue area mainly for items people are currently working on a PR for? I probably won't work on that any time soon.

(We only set canned acls to backport iceberg tables from one account to hive tables in a legacy account, so my team doesn't need this in order to move onto S3FileIO)

jackye1995 replied:
Yes I think canned ACL is a feature for another PR, and I was planning to do the PR after #1754 , let's wait until that is checked in so that we have a stable interface for adding new things.

danielcweeks replied:
I'd say wait until you feel like you need it and then file an issue.  Filing the issue doesn't mean you have to work on it, but this feels like one of the many features of the client that may not be used very much.  It should be trivial to add, so if you find that you do need it, we can easily add it then.

jackye1995 replied:
Yes. This is one of the things I planned to port, I am fine either me or @johnclara do it.

johnclara replied:
@jackye1995 if you were already planning on it, go ahead! :) my team doesn't need it any time soon but it would be good for people who run into a similar issue

johnclara commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -69,14 +132,54 @@ public void flush() throws IOException {
+      stream.close();
+    }
+
+    currentStagingFile = File.createTempFile("s3fileio-", ".tmp");
```
Not sure if the staging directory should be configurable at some point (not this PR)?

For our clients in docker containers, we've been thinking about moving s3a's staging area to be under a mount optimized for IO.

We haven't done any tests to see if this gives us any noticeable performance gains.

jackye1995 replied:
I think you can achieve that by configuring an alternative temp directory when starting the JVM, is this an option for you?

danielcweeks replied:
Yeah, this makes sense and should be easy to add once we have properties pushed all the way down from the catalog/tables (you'll see that mentioned about in https://github.com/apache/iceberg/pull/1767#discussion_r523154834).  You might get some savings by using st1 volumes for this specifically due to it being entirely sequential read and write, so that might be interesting to experiment with.

danielcweeks replied:
P.S. I would also suggest that as a follow on issue.

danielcweeks commented:
@jackye1995 I've moved this PR to draft for now.  I rebased on top of your changes in #1754, but it's somewhat complicated due to the fact that now we have 3 separate requests (create multipart, upload part, put object) that all require setting the properties, but they don't inherit from a common interface.  I feel like reflection is probably the best way do simplify this, but I'll push the current state of things.

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:
```diff
@@ -19,42 +19,83 @@
 class S3OutputStream extends PositionOutputStream {
   private static final Logger LOG = LoggerFactory.getLogger(S3OutputStream.class);
 
+  static final String MULTIPART_SIZE = "s3fileio.multipart.size";
```
Converted to use AwsProperties

jackye1995 commented:
> @jackye1995 I've moved this PR to draft for now. I rebased on top of your changes in #1754, but it's somewhat complicated due to the fact that now we have 3 separate requests (create multipart, upload part, put object) that all require setting the properties, but they don't inherit from a common interface. I feel like reflection is probably the best way do simplify this, but I'll push the current state of things.

Sorry I was dealing with some other errands yesterday. Please see #1786 to see if that solves this issue, I am trying to not use reflection, but we can also switch to that approach if it is simpler.

danielcweeks commented:
> > @jackye1995 I've moved this PR to draft for now. I rebased on top of your changes in #1754, but it's somewhat complicated due to the fact that now we have 3 separate requests (create multipart, upload part, put object) that all require setting the properties, but they don't inherit from a common interface. I feel like reflection is probably the best way do simplify this, but I'll push the current state of things.
> 
> Sorry I was dealing with some other errands yesterday. Please see #1786 to see if that solves this issue, I am trying to not use reflection, but we can also switch to that approach if it is simpler.

Turns out I did something very similar here: https://github.com/apache/iceberg/pull/1767/files#diff-133c36e9cbb025f7cb44c2daac330c501d85a5a6eb44126c0eb6155f2bad7407R30

jackye1995 commented:
> > > @jackye1995 I've moved this PR to draft for now. I rebased on top of your changes in #1754, but it's somewhat complicated due to the fact that now we have 3 separate requests (create multipart, upload part, put object) that all require setting the properties, but they don't inherit from a common interface. I feel like reflection is probably the best way do simplify this, but I'll push the current state of things.
> > 
> > 
> > Sorry I was dealing with some other errands yesterday. Please see #1786 to see if that solves this issue, I am trying to not use reflection, but we can also switch to that approach if it is simpler.
> 
> Turns out I did something very similar here: https://github.com/apache/iceberg/pull/1767/files#diff-133c36e9cbb025f7cb44c2daac330c501d85a5a6eb44126c0eb6155f2bad7407R30

Oh yes, looks like It is a generalization for that class. For the name, I would prefer the class to be more generic just to avoid creating other utils for future use cases. For reflection, I am not sure what is the community guideline here but personally I would avoid using that, which resulted in that solution using functional interface.

danielcweeks commented:
> > > > @jackye1995 I've moved this PR to draft for now. I rebased on top of your changes in #1754, but it's somewhat complicated due to the fact that now we have 3 separate requests (create multipart, upload part, put object) that all require setting the properties, but they don't inherit from a common interface. I feel like reflection is probably the best way do simplify this, but I'll push the current state of things.
> > > 
> > > 
> > > Sorry I was dealing with some other errands yesterday. Please see #1786 to see if that solves this issue, I am trying to not use reflection, but we can also switch to that approach if it is simpler.
> > 
> > 
> > Turns out I did something very similar here: https://github.com/apache/iceberg/pull/1767/files#diff-133c36e9cbb025f7cb44c2daac330c501d85a5a6eb44126c0eb6155f2bad7407R30
> 
> Oh yes, looks like It is a generalization for that class. For the name, I would prefer the class to be more generic just to avoid creating other utils for future use cases. For reflection, I am not sure what is the community guideline here but personally I would avoid using that, which resulted in that solution using functional interface.

@jackye1995 I actually stepped back from using reflection as I found out that some request types do not set all of the same parameters (e.g. UploadPartRequest, GetObjectRequest aren't exactly the same as PutObjectRequest and CreateMultipartUploadRequest).  Since there are no common interfaces, it seems reasonable to just separate out the behavior for now.

As we get into other instances of this (like ACL, etc.), then maybe it would make sense to find a more concise solution.

I've got a few small changes to make and want to expand the testing, but hopefully have something ready for review today.

jackye1995 commented:
The ACL support #1788  should not be a big issue, it is only a single additional method call `acl(...)` and only affects `PutObjectRequest` and `CreateMultipartUploadRequest`. 

I was thinking there is still a way to use reflection, using a similar logic as #1786 . We need to check if the method exists or not when setting the value. The biggest concern I have with reflection is that I don't know how much it affects the performance since it is in the critical path for upload. Let me run some tests and come back with some data for that.

rdblue reviewed (approved):
Looks good to me!
